applications/OptimizationApplication/custom_utilities/container_variable_data_holder/container_variable_data_holder.cpp

```diff
@@ -0,0 +1,657 @@
+
+    auto& r_data = this->mData;
+    auto& r_container = this->GetContainer();
+
+    r_data.resize(number_of_entities * this->mDataDimension);
+
+    IndexPartition<IndexType>(number_of_entities).for_each([&](const IndexType Index){
+        const auto& values = TContainerDataIO::GetValue(*(r_container.begin() + Index), rVariable);
+        for (IndexType i = 0; i < this->mDataDimension; ++i) {
+            ContainerVariableDataHolderHelperUtilities::AssignValueToVector(r_data, Index * this->mDataDimension, i, values);
+        }
+    });
+
+    KRATOS_CATCH("")
+}
+
+template <class TContainerType, class TContainerDataIO>
+template<class TDataType>
+void ContainerVariableDataHolder<TContainerType, TContainerDataIO>::AssignDataToContainerVariable(const Variable<TDataType>& rVariable)
+{
+    KRATOS_TRY
+
+    KRATOS_ERROR_IF_NOT(this->GetModelPart().GetProcessInfo().Has(DOMAIN_SIZE))
+        << "DOMAIN_SIZE variable is not found in process info of model part "
+           "with name "
+        << this->GetModelPart().FullName() << ".\n";
+
+    const IndexType local_size = ContainerVariableDataHolderHelperUtilities::GetLocalSize<TDataType>(this->GetModelPart().GetProcessInfo()[DOMAIN_SIZE]);
+
+    KRATOS_ERROR_IF_NOT(local_size == this->GetDataDimension())
+        << "Stored data dimension and requested assignment variable data "
+           "dimension mismatch. [ Sotred data dimension = "
+        << this->GetDataDimension()
+        << ", assignment requested variable data diemension = " << local_size << " ].\n";
+
+    const IndexType number_of_entities = this->GetContainer().size();
+
+    const auto& r_values = this->mData;
+    auto& r_container = this->GetContainer();
+
+    KRATOS_ERROR_IF_NOT(number_of_entities * local_size == r_values.size())
+        << "Container size and values size mismatch. [ Values size = "
+        << r_values.size() << ", container size = " << number_of_entities
+        << ", local size = " << local_size << " ].\n";
+
+    // initialize the container variables first
+    if constexpr(std::is_same_v<TContainerType, ModelPart::NodesContainerType>) {
+        // initializes ghost nodes as well for the later synchronization
+        // only, the nodal non historical values needs to be set unless
+        // they are properly initialized.
+        block_for_each(this->GetModelPart().Nodes(), [&](auto& rEntity) {
+            TContainerDataIO::SetValue(rEntity, rVariable, rVariable.Zero());
+        });
+    } else {
+        // no ghost settings required for elements or conditions nor historical values, hence no special treatment required.
+        block_for_each(r_container, [&](auto& rEntity) {
+            TContainerDataIO::SetValue(rEntity, rVariable, rVariable.Zero());
+        });
+    }
+
+    IndexPartition<IndexType>(number_of_entities).for_each([&](const IndexType Index){
+        auto& values = TContainerDataIO::GetValue(*(r_container.begin() + Index), rVariable);
```

Comment:
naming

there's `r_values` and there's `values`, both of which are references and aren't too descriptive. How about `r_source_values` and `r_target_values`?